### PR TITLE
Fix: Improve Elixir highlighting performance

### PR DESCRIPTION
**Pull request by connorlay**
Resolves https://github.com/nvim-treesitter/nvim-treesitter/issues/1486

This PR rewrites the highlight queries for Elixir with a focus on performance. I based these new highlights on prior work: https://github.com/ananthakumaran/tree-sitter-elixir/blob/master/queries/highlights.scm and https://github.com/nvim-treesitter/nvim-treesitter/blob/master/queries/elixir/highlights.scm. On my machine this branch makes things much more usable, but I don't have exact numbers to report. Is there a way to benchmark these highlights?

Additionally, these new highlights only consider the core language and ignore commonly used libraries like Ecto, Phoenix, etc. Does it make sense to add in highlights for these libraries, or leave that up to the user to customize?

### CODE REVIEW

**Comment by nifoc** [comment]
Thank you for simplifying the queries!

I only took a quick glance at the changes. Will try them out tomorrow.

Personally I'm not a fan of loosing so many specific highlights (conditionals for example), but if even those are too slow on some machines I'm obviously not opposed.

Btw. should the other queries be simplified, too?

**Comment by connorlay** [comment]
> Personally I'm not a fan of loosing so many specific highlights (conditionals for example), but if even those are too slow on some machines I'm obviously not opposed.

Thanks for the feedback! I went ahead and added back in conditionals, repeat, and exceptions and (on my machine) I don't see a negative performance impact. I agree it reads a bit better 👀

**Comment by connorlay** [comment]
> Btw. should the other queries be simplified, too? (locals are quite complex)

I haven't used the locals plugin much myself, but I am happy to look into it as part of a separate PR.

**Comment by nifoc** [comment]
> I haven't used the locals plugin much myself, but I am happy to look into it as part of a separate PR.

I use them via nvim-compe and they don't feel slow to me, but then again I don't notice the highlight stuff either …

**Comment by nifoc** [comment]
I played around with this is a bit today and it looks good to me at the moment 👍

**Comment by Yarnus** [comment]
I did some simple tests on my machine and use below snippet:

```elixir
defmodule Demo do
  @moduledoc false
end
```

- by master commit, it's always around 425 milliseconds

```log
340.189  000.209: before starting main loop
768.000  427.811: first screen update
768.004  000.004: --- NVIM STARTED ---

225.091  000.222: before starting main loop
650.702  425.611: first screen update
650.705  000.003: --- NVIM STARTED ---

229.387  000.193: before starting main loop
654.070  424.683: first screen update
654.076  000.006: --- NVIM STARTED ---
```

- by connorlay's PR, It's down to about 150 milliseconds
```log
234.560  000.201: before starting main loop
389.290  154.730: first screen update
389.294  000.004: --- NVIM STARTED ---

253.941  000.249: before starting main loop
407.982  154.040: first screen update
407.985  000.003: --- NVIM STARTED ---

226.624  000.197: before starting main loop
383.197  156.573: first screen update
383.201  000.003: --- NVIM STARTED ---
```
it seems much better than before.  Nice job !

File: queries/elixir/highlights.scm [hunk 1]
```diff
@@ -1,180 +1,154 @@
...
-  (sigil_end)
   (sigil_start)
-] @string.special
+  (sigil_end)
+  (heredoc_start)
+  (heredoc_end)
+] @punctuation.special
 
-(escape_sequence) @string.escape
+(interpolation "#{" @punctuation.special "}" @punctuation.special)
```
Comment (vigoux) [inline queries/elixir/highlights.scm:27]: Personal opinion: I rather like the following, does it affect performances ?
```suggestion
(interpolation ["#{" "}"] @punctuation.special)
```

File: queries/elixir/highlights.scm [hunk 2]
```diff
@@ -1,180 +1,154 @@
...
+] @string
 
-; Call to a local function
-(call (function_identifier) @method)
+; __MODULE__ and friends
+(special_identifier) @constant.builtin
+
+(map "%{" @constructor "}" @constructor)
+
+(struct "%" @constructor "{" @constructor "}" @constructor)
```
Comment (vigoux) [inline queries/elixir/highlights.scm:61]: Same as above, maybe use alternations (things in `[]`) here too?

File: queries/elixir/highlights.scm [hunk 3]
```diff
@@ -1,180 +1,154 @@
...
+
+(binary_op operator: _ @operator)
```
Comment (vigoux) [inline queries/elixir/highlights.scm:63]: We tend to put parens around wildcard nodes :
```suggestion
(binary_op operator: (_) @operator)
```

Comment (connorlay) [reply queries/elixir/highlights.scm:63]: Hmm, I tried your suggestion and that broke highlighting for unary and binary operators 😦 Could we move forward with this as-is for now and revisit this later?

**Comment by connorlay** [comment]
@vigoux Thank you for the feedback! I pushed a change with your suggestions applied.

File: queries/elixir/highlights.scm (hunk 3 above)
Comment (stsewd) [reply queries/elixir/highlights.scm:63]: `_` matches both named and anonymous nodes, `(_)` matches named nodes only.

Comment (connorlay) [reply queries/elixir/highlights.scm:63]: That would explain why the highlights broke with `(_)`.  Based on what I see in the playground, `operator` is an anonymous Elixir node that you can reference via a field.

**Review by vigoux — APPROVED** [review]

**Comment by jherdman** [comment]
Will this be back ported to the 0.5-compat branch?

**Comment by stsewd** [comment]
> Will this be back ported to the 0.5-compat branch?

Currently that branch is 93 commit behind, I think the only breaking change that has happened is https://github.com/nvim-treesitter/nvim-treesitter/pull/1626, so it should be safe to port, but I'm not sure how we ended up in the procedure to do so. @vigoux may have more information here.

**Pull request merged**